code/game/machinery/deployable.dm

```diff
@@ -210,6 +218,167 @@
+	stacktype = null
+	/// This variable is used to tell the shield to ping it's owner when it is broke.
+	var/core_shield = FALSE
+	/// This variable is to tell the shield what it's source is.
+	var/obj/structure/dropwall_generator/source = null
+	explosion_block = 8 //should be enough for a potasium water nade that isn't a maxcap. If you stand next to a maxcap with this however, it will end poorly
+
+/obj/structure/barricade/dropwall/Initialize(mapload, owner, core, dir_1, dir_2)
+	. = ..()
+	source = owner
+	core_shield = core
+	directional_list += dir_1
+	directional_list += dir_2
+	if(dir_2)
+		icon_state = "[dir2text(dir_1 + dir_2)]"
+	else
+		icon_state = "[dir2text(dir_1)]"
+
+/obj/structure/barricade/dropwall/Destroy()
+	if(core_shield)
+		source.protected = FALSE
+	return ..()
+
+/obj/structure/barricade/dropwall/emp_act(severity)
+	..()
+	take_damage( 40 / severity, BRUTE) //chances are the EMP will also hit the generator, we don't want it to double up too heavily
+
+/obj/structure/barricade/dropwall/bullet_act(obj/item/projectile/P)
+	if(P.shield_buster)
+		qdel(src)
+	else
+		. = ..()
+
+/obj/item/grenade/barrier/dropwall
+	name = "dropwall shield generator"
+	desc = "This generator designed by Shellguard Munitions's spartan division is used to deploy a temporary cover that blocks projectiles and explosions from a direction, while allowing projectiles to pass freely from behind."
+	actions_types = list(/datum/action/item_action/toggle_barrier_spread)
+	icon = 'icons/obj/dropwall.dmi'
+	icon_state = "dropwall"
+	mode = NORTH
+	var/uptime = 12 SECONDS
+
+/obj/item/grenade/barrier/dropwall/toggle_mode(mob/user)
+	switch(mode)
+		if(NORTH)
+			mode = EAST
+		if(EAST)
+			mode = SOUTH
+		if(SOUTH)
+			mode = WEST
+		if(WEST)
+			mode = NORTH
+
+	to_chat(user, "[src] is now in [dir2text(mode)] mode.")
+
+/obj/item/grenade/barrier/dropwall/prime()
+	new /obj/structure/dropwall_generator(get_turf(loc), mode, uptime)
+	qdel(src)
+
+/obj/structure/dropwall_generator
+	name = "deployed dropwall shield generator"
+	desc = "This generator designed by Shellguard Munitions's spartan division is used to deploy a temporary cover that blocks projectiles and explosions from a direction, while allowing projectiles to pass freely from behind."
+	icon = 'icons/obj/dropwall.dmi'
+	icon_state = "dropwall_deployed"
+	max_integrity = 25 // 2 shots
+	var/list/connected_shields = list()
+	/// This variable is used to prevent damage to it's core shield when it is up.
+	var/protected = FALSE
+	///The core shield that protects the generator
+	var/obj/structure/barricade/dropwall/core_shield = null
+
+/obj/structure/dropwall_generator/Initialize(mapload, direction, uptime)
+	. = ..()
+	if(direction)
+		deploy(direction, uptime)
+
+/obj/structure/dropwall_generator/proc/deploy(direction, uptime)
+	anchored = TRUE
+	protected = TRUE
+	addtimer(CALLBACK(src, .proc/power_out), uptime)
+	while(uptime >= 10)
```

Comment:
Took me a while to realise this is done in a while loop. It might be cleaner to add a new timer when the old timer goes off.
Also removes the need to explain what is going on here with the calculations.
Just be sure to update your logic as well since you divide by 10 here

Reply:
Reworked to be a proc that calls itself.